// mappersmith/gateway/mock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _gateway = _interopRequireDefault(require("../gateway"));

var _test = require("../test");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Mock(request) {
  _gateway.default.apply(this, arguments);
}

Mock.prototype = _gateway.default.extends({
  get: function get() {
    this.callMock();
  },
  head: function head() {
    this.callMock();
  },
  post: function post() {
    this.callMock();
  },
  put: function put() {
    this.callMock();
  },
  patch: function patch() {
    this.callMock();
  },
  delete: function _delete() {
    this.callMock();
  },
  callMock: function callMock(httpMethod) {
    var _this = this;

    return (0, _test.lookupResponseAsync)(this.request).then(function (response) {
      return _this.dispatchResponse(response);
    }).catch(function (e) {
      return _this.dispatchClientError(e.message, e);
    });
  }
});
var _default = Mock;
exports.default = _default;
